bluemira/builders/tools.py

```diff
@@ -118,62 +118,39 @@ def circular_pattern_component(
-
-        comp = copy.deepcopy(comp)
-        comp.name = f"{comp.name} {idx}"
-
-        comp.children = []
-        orig_parent: bm_comp.Component = comp.parent
-        if orig_parent is not None:
-            comp.parent = sector.get_component(f"{orig_parent.name} {idx}")
-        if comp.parent is None:
-            comp.parent = sector
+    component = [component] if isinstance(component, bm_comp.Component) else component
+    sectors = [bm_comp.Component(f"{parent_prefix}") for _ in range(n_children)]
+    # build sector trees by assigning copies of each component to sec. parents
+    for c in component:
+        for parent_sc in sectors:
+            c.copy(parent_sc)
 
+    sector_tree_indexs = [list(PreOrderIter(sc)) for sc in sectors]
+
+    # to keep naming convention
+    for sec_i, sector_index in enumerate(sector_tree_indexs):
+        for comp in sector_index:
+            comp.name = f"{comp.name} {sec_i + 1}"
+
+    faux_sec_comp = bm_comp.Component(f"{parent_prefix} X")
+    faux_sec_comp.children = component
+
+    for search_index_i, comp in enumerate(PreOrderIter(faux_sec_comp)):
         if isinstance(comp, bm_comp.PhysicalComponent):
-            comp.shape = shape
-
-    def assign_or_pattern(comps: Union[bm_comp.Component, List[bm_comp.Component]]):
-        if not isinstance(comps, list):
-            comps = [comps]
-
-        for comp in comps:
-            if isinstance(comp, bm_comp.PhysicalComponent):
-                shapes = bm_geo.tools.circular_pattern(
-                    comp.shape,
-                    n_shapes=n_children,
-                    origin=origin,
-                    direction=direction,
-                    degree=degree,
-                )
-                for sector, shape in zip(sectors, shapes):
-                    assign_component_to_sector(comp, sector, shape)
-            else:
-                for sector in sectors:
-                    assign_component_to_sector(comp, sector)
-
-    def process_children(comps: Union[bm_comp.Component, List[bm_comp.Component]]):
-        if not isinstance(comps, list):
-            comps = [comps]
-        for comp in comps:
-            for child in comp.children:
-                assign_or_pattern(child)
-                process_children(child)
-
-    assign_or_pattern(component)
-    process_children(component)
+            shapes = bm_geo.tools.circular_pattern(
+                comp.shape,
+                n_shapes=n_children,
+                origin=origin,
+                direction=direction,
+                degree=degree,
+            )
+            # assign each shape to each sector at index search_index_i
+            # which should be the copy of the PhysicalComponent
+            for sector_index, shape in zip(sector_tree_indexs, shapes):
+                phy_comp = sector_index[search_index_i]
+                if not isinstance(phy_comp, bm_comp.PhysicalComponent):
+                    raise ComponentError("Something has gone wrong :/")
```

Comment:
Any more explanation for this error? When might we expect "something to go wrong"?

Reply:
The expectation should really never be raised. But ya I can make it more clear